pyscf/sfnoci/sfnoci.py

```diff
@@ -0,0 +1,1266 @@
+    '''SFNOCI solver
+
+    Args:
+        sfnoci: SFNOCI object
+
+        mo_coeff : ndarray
+            orbitals to construct active space Hamiltonian
+        ci0 : ndarray or custom types
+            FCI sovler initial guess. For external FCI-like solvers, it can be
+            overloaded different data type. For example, in the state-average
+            FCI solver, ci0 is a list of ndarray. In other solvers such as
+            DMRGCI solver, SHCI solver, ci0 are custom types.
+
+    kwargs:
+        envs: dict
+            The variable envs is created (for PR 807) to passes MCSCF runtime
+            environment variables to SHCI solver. For solvers which do not
+            need this parameter, a kwargs should be created in kernel method
+            and "envs" pop in kernel function
+    '''
+    if mo_coeff is None: mo_coeff = sfnoci.mo_coeff
+    if ci0 is None: ci0 = sfnoci.ci
+
+    log = logger.new_logger(sfnoci, verbose)
+    t0 = (logger.process_clock(), logger.perf_counter())
+    if hasattr(sfnoci, '_groupA'): log.debug('Start SFGNOCI')
+    else: log.debug('Start SFNOCI')
+
+
+    ncas = sfnoci.ncas
+    nelecas = sfnoci.nelecas
+
+    # FASSCF
+    mo_list, po_list, group = sfnoci.optimize_mo(mo_coeff)
+    conf_info_list = group_info_list(ncas, nelecas, po_list, group)
+    t1 = log.timer('FASSCF', *t0)
+
+    # SVD and core density matrix
+    if group is None:
+        dmet_core_list, ov_list = sfnoci.get_svd_matrices(mo_list, po_list)
+    else:
+        dmet_core_list, ov_list = sfnoci.get_svd_matrices(mo_list, group)
+    t1 = log.timer('SVD and core density matrix', *t1)
+
+    # 1e
+    dmet_act_list = sfnoci.get_active_dm(mo_coeff)
+    h1e, ecore_list = sfnoci.get_h1cas(dmet_act_list , mo_list , dmet_core_list)
+    t1 = log.timer('effective 1e hamiltonians and core energies', *t1)
+
+    # 2e
+    eri = sfnoci.get_h2eff(mo_coeff)
+    t1 = log.timer('effective 2e hamiltonian', *t1)
+
+    # FCI
+    max_memory = max(400, sfnoci.max_memory-lib.current_memory()[0])
+    e_tot, fcivec = sfnoci.fcisolver.kernel(h1e, eri, ncas, nelecas,
+                                            conf_info_list, ov_list, ecore_list,
+                                            ci0=ci0, verbose=log,
+                                            max_memory=max_memory)
+
+    log.timer('SFNOCI solver', *t1)
+    log.timer('All SFNOCI process', *t0)
+
+    if isinstance(e_tot, (float, numpy.float64)):
+        e_cas = e_tot - ecore_list
+    else:
+        e_cas = [e - ecore_list for e in e_tot]
+
+    return e_tot, e_cas, fcivec
+
+def possible_occ(n_as,n_ase):
+    def find_arrays(n_as,n_ase):
+        possible_values=[0,1,2]
+        arrays=[]
+
+        for combination in product(possible_values,repeat=n_as):
+            if sum(combination)==n_ase:
+                arrays.append(numpy.array(combination))
+        return arrays
+    result_arrays=find_arrays(n_as,n_ase)
+    array_list=[]
```

Comment:
why not directly calling
`concatenated_array = np.array(result_arrays, dtype=np.int32)` ?